### PR TITLE
Refactor of the formatter to scale better later on

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
This is a proposition so it's easier to add Formatter later on and remove some duplicated calls. (Based on #8)

### CODE REVIEW

**Comment by DanielNoord** [comment]
I like this approach!

File: pydocstringformatter/formatting.py [hunk 1]
```diff
@@ -1,39 +1,74 @@
+import abc
 import re
 import tokenize
 
+# pylint: disable=too-few-public-methods
```
Comment (DanielNoord) [inline pydocstringformatter/formatting.py:5]: Could you add this to `pyproject.toml`? I keep ignoring this in my projects anyway 😄

File: pydocstringformatter/formatting.py [hunk 2]
```diff
@@ -1,39 +1,74 @@
...
+from typing import List
 
-def _format_beginning_quotes(tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
+
+class Formatter:
+    """Base class for docstring formatter"""
+
+    @abc.abstractmethod
+    def treat(self, tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
```
Comment (DanielNoord) [inline pydocstringformatter/formatting.py:13]:
```suggestion
    def treat_token(self, tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
```

File: pydocstringformatter/formatting.py [hunk 3]
```diff
@@ -1,39 +1,74 @@
...
+        """Return a modified token"""
+
+
+class StringFormatter(Formatter):
+    """Base class for formatter that only modify the string content"""
```
Comment (DanielNoord) [inline pydocstringformatter/formatting.py:18]:
```suggestion
    """Base class for formatter that only modifies the string content"""
```

File: pydocstringformatter/formatting.py [hunk 4]
```diff
@@ -1,39 +1,74 @@
...
+
+    @abc.abstractmethod
+    def _treat_string(self, new_string: str) -> str:
+        """Will modify the string"""
+
+    def treat(self, tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
```
Comment (DanielNoord) [inline pydocstringformatter/formatting.py:24]:
```suggestion
    def treat_token(self, tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
```

File: pydocstringformatter/formatting.py [hunk 5]
```diff
@@ -1,39 +1,74 @@
...
-
-def _format_closing_quotes(tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
+
+    def _treat_string(self, new_string: str) -> str:
+        if new_string[3] == "\n":
+            new_string = re.sub(r"\n *", "", new_string, 1)
+        return new_string
+
+
+class ClosingQuotesFormatter(Formatter):
```
Comment (DanielNoord) [inline pydocstringformatter/formatting.py:43]: Shouldn't this be a `StringFormatter` as well?

File: pydocstringformatter/formatting.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply pydocstringformatter/formatting.py]: Done

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Just something I found while reading the PEPs (and to be added at a later point) one-liner's and summary lines should be sentences ending with a point. I have not done that myself for some of the docstrings I already wrote, but we might as well start doing this ourselves now.

File: pydocstringformatter/formatting.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply pydocstringformatter/formatting.py]: Done (We should make this one optional asap, it's very annoying to remove it first thing in all new projects 😄)

File: pydocstringformatter/formatting.py (hunk 5 above)
Comment (Pierre-Sassoulas) [reply pydocstringformatter/formatting.py]: ``return self._format_multiline_ending_quotes(tokeninfo)`` returns a token

Comment (DanielNoord) [reply pydocstringformatter/formatting.py]: It does, but only ever changes the `string`. It's a bad design that gets shown by this wonderful refactor! 😄

**Comment by Pierre-Sassoulas** [comment]
I can remove the ``setup.py`` commit so we merge this one first if this is controversial 😉 .

**Comment by DanielNoord** [comment]
> I can remove the `setup.py` commit so we merge this one first if this is controversial 😉 .

Would be good I think! I hate having to scroll down two pages on a Github repository because of all the config files, that's why I'd like to enforce not having a `setup.py`.

I should add a `build-system` section though, so that's on the TODO-list.

File: pydocstringformatter/formatting.py (hunk 5 above)
Comment (Pierre-Sassoulas) [reply pydocstringformatter/formatting.py]: Hmm but it needs the token and not the string to work. Maybe treat_string needs to take a token as input ?

Comment (Pierre-Sassoulas) [reply pydocstringformatter/formatting.py]: Or we could initialize the formatters for each loop and the token become a class attribute.

Comment (DanielNoord) [reply pydocstringformatter/formatting.py]: Ah that makes sense I think.

Comment (DanielNoord) [reply pydocstringformatter/formatting.py]: > Or we could initialize the formatters for each loop and the token become a class attribute.

That sounds like a large impact on performance. Would probably still be manageable, but I like how quick the tool runs at the moment.

Comment (Pierre-Sassoulas) [reply pydocstringformatter/formatting.py]: I've refactored treat_string, the result does look cleaner.

File: pydocstringformatter/formatting/__init__.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+from typing import List
+
+from pydocstringformatter.formatting.formatter import (
+    BeginningQuotesFormatter,
+    ClosingQuotesFormatter,
+    Formatter,
+)
+
+FORMATTERS: List[Formatter] = [
```
Comment (DanielNoord) [inline pydocstringformatter/formatting/__init__.py:11]:
```suggestion
FORMATTERS = [
```

File: pydocstringformatter/formatting/__init__.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+__all__ = ["FORMATTERS"]
+
+from typing import List
```
Comment (DanielNoord) [inline pydocstringformatter/formatting/__init__.py:3]:
```suggestion
```

File: pydocstringformatter/formatting/formatter.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    def treat_token(self, tokeninfo: tokenize.TokenInfo) -> tokenize.TokenInfo:
+        """Return a modified token"""
+
+
+class StringFormatter(Formatter):
+    """Base class for formatter that only modifies the string content"""
+
+    @abc.abstractmethod
+    def _treat_string(self, tokeninfo: tokenize.TokenInfo) -> str:
+        """Will modifies the string"""
```
Comment (DanielNoord) [inline pydocstringformatter/formatting/formatter.py:19]:
```suggestion
        """Return a modified string"""
```

**Review by DanielNoord — APPROVED** [review]
Three final comments! I think this can be squashed and merged 😄 

Congratulations on becoming a `pydocstringformatter` contributor 😉

**Comment by Pierre-Sassoulas** [comment]
> Congratulations on becoming a pydocstringformatter contributor

It's actually nice to be on a project with a bug to feature ratio of 0%, how do you do that ? 😄

**Comment by DanielNoord** [comment]
> > Congratulations on becoming a pydocstringformatter contributor
> 
> It's actually nice to be on a project with a bug to feature ratio of 0%, how do you do that ? 😄

Rigorous reviews of all PRs 😄

**Pull request merged**